### PR TITLE
Improve ethminer perf when using --cl-ms-per-batch 

### PR DESCRIPTION
**Pull request by bargst**

### CODE REVIEW

File: libethash-cl/ethash_cl_miner.cpp [hunk 1]
```diff
@@ -520,14 +522,26 @@ void ethash_cl_miner::search(uint8_t const* header, uint64_t target, search_hook
...
+						// cerr << "Batch of " << m_globalWorkSize << " took " << chrono::duration_cast<chrono::milliseconds>(d).count() << " ms, >> " << s_msPerBatch << " ms." << endl;
+						m_globalWorkSize = max<unsigned>(128, m_globalWorkSize - m_stepWorkSizeAdjust);
 						// cerr << "New global work size" << m_globalWorkSize << endl;
 					}
 					else if (d < chrono::milliseconds(s_msPerBatch * 9 / 10))
 					{
-						// cerr << "Batch of " << m_globalWorkSize << " took " << chrono::duration_cast<chrono::milliseconds>(d).count() << " ms, << " << _msPerBatch << " ms." << endl;
-						m_globalWorkSize = min<unsigned>(pow(2, m_deviceBits) - 1, m_globalWorkSize - s_workgroupSize);
+						// Divide the step by 2 when adjustment way change
+						if (m_wayWorkSizeAdjust < 1 )
```
Comment (gavofyork) [inline libethash-cl/ethash_cl_miner.cpp]: space before `)`

File: libethash-cl/ethash_cl_miner.cpp [hunk 2]
```diff
@@ -520,14 +522,26 @@ void ethash_cl_miner::search(uint8_t const* header, uint64_t target, search_hook
 				{
 					if (d > chrono::milliseconds(s_msPerBatch * 10 / 9))
 					{
-						// cerr << "Batch of " << m_globalWorkSize << " took " << chrono::duration_cast<chrono::milliseconds>(d).count() << " ms, >> " << _msPerBatch << " ms." << endl;
-						m_globalWorkSize = max<unsigned>(128, m_globalWorkSize + s_workgroupSize);
+						// Divide the step by 2 when adjustment way change
+						if (m_wayWorkSizeAdjust > -1 )
```
Comment (gavofyork) [inline libethash-cl/ethash_cl_miner.cpp]: space before `)`

File: libethash-cl/ethash_cl_miner.cpp [hunk 3]
```diff
@@ -315,6 +315,8 @@ bool ethash_cl_miner::init(
 			m_globalWorkSize = ((m_globalWorkSize / s_workgroupSize) + 1) * s_workgroupSize;
 		// remember the device's address bits
 		m_deviceBits = device.getInfo<CL_DEVICE_ADDRESS_BITS>();
+		// make sure first step of global work size adjustment is large enough
+		m_stepWorkSizeAdjust = pow(2, m_deviceBits/2+1);
```
Comment (gavofyork) [inline libethash-cl/ethash_cl_miner.cpp]: insert spaces in arithmetic.

**Comment by gavofyork** [comment]
what sort of performance increase does this give?

@LefterisJP merge when style done and you're happy with it.

**Comment by bargst** [comment]
Perf increase for an AMD R9 280 : from 20.1MH/s to 21.4MH/s when using --cl-ms-per-batch 50

**Comment by LefterisJP** [comment]
For me I could not see performance increase but this adds nothing but some more tweaking so I don't see anything bad, especially if it helps with some of our user's hardware and `--cl-ms-per-batch`

**Pull request merged**